#**EDA**

#**EMAX:**

José Emiliano Díaz Gutiérrez | A01711141

Ximena Cantera Reséndiz       | A01277310

Ma. Fernanda Barba De Los Santos  | A01657868


#**INTRODUCCIÓN**
En este notebook, realizaremos un análisis de diferentes archivos que nos proporcionaron CNH relacionados con las unidades que utilizan, su actividad, mantenimiento, rutas y rankings.

El objetivo principal es integrar estos archivos, estandarizar sus formatos, limpiar los datos, y construir nuevas características que permitan una comprensión 360° del ciclo de vida y desempeño de cada unidad.

Para finalizar el análisis y contrucción, aplicaremos técnicas de codificación y transformación para preparar los datos para futuros modelos de análisis y toma de decisiones.

# **Análisis Exploratorio de Datos (EDA)**


## 1. Carga de Datos Inicial y Vistas Previas

In [4]:
import pandas as pd

df_top50 = pd.read_excel("../data/ArchivosLimpios/new_distribuidor.xlsx")
df_horas = pd.read_excel("../data/ArchivosLimpios/new_horas.xlsx")
df_rutas = pd.read_excel("../data/ArchivosLimpios/new_ruteo.xlsx")
df_mantenimiento = pd.read_excel("../data/ArchivosLimpios/new_mantenimientos.xlsx")
df_population = pd.read_excel("../data/ArchivosLimpios/new_population.xlsx")
df_unidades = pd.read_excel("../data/ArchivosLimpios/new_unidades.xlsx", header=1)

# **PopulationView_2026.xlsx**

In [5]:
print("PopulationView_2026.xlsx")
df_population.head()

PopulationView_2026.xlsx


,VIN (17 CHARACTERS),VIN,BRAND,TYPE,PRODUCT,SERIES,MODEL,TECHNICAL TYPE,ICE CODE,DEVICE SERIAL NUMBER,...,TELEMATICS BRAND,TELEMATICS TYPE,TELEMATICS MODEL,OWNER NAME,ENGINE HOURS,LATEST COMMUNICATION DATE,BDE PROCESSED AT TIME,MYFLEET DEALER,MYFLEET DEALER NAME,EULA ACCEPTED
0,NaN,FR1702510,CASE IH,TRACTORS,UTILITY,FARMALL® JXM,Farmall® 80JXM,696055307,A.A.02.013.004.003,1602-38000E07,...,CS,TRACTORS,FRMLL 80JXM,"Maquinas Diesel Sa De Cv, Santa Catarina, Nuev...",10,2026-02-08 18:55:15,2026-02-09 00:33:33,190-08B114A,Maquinas Diesel Sa De Cv,Desconocido
1,NaN,FR1702521,CASE IH,TRACTORS,UTILITY,FARMALL® JXM,Farmall® 80JXM,696055307,A.A.02.013.004.003,1602-38000BD1,...,CS,TRACTORS,FRMLL 80JXM,"Maquinas Diesel Sa De Cv, Santa Catarina, Nuev...",1,2026-02-03 12:38:49,2026-02-09 00:33:33,190-08B114A,Maquinas Diesel Sa De Cv,Desconocido
2,NaN,FR1706211,CASE IH,TRACTORS,UTILITY,FARMALL® JXM,Farmall® 80JXM,696055308,A.A.02.013.004.004,1602-38000C1E,...,CS,TRACTORS,CM80,Acuicola Tacachi,1,2026-02-08 12:21:20,2026-02-08 23:33:33,190-08B059A,Alfonso R. Bours S.A. De C.V.,Yes
3,NaN,FR1706358,CASE IH,TRACTORS,UTILITY,FARMALL® JXM,Farmall® 90JXM,696055310,A.A.02.013.005.004,1602-38000BB7,...,CS,TRACTORS,CM90,"Agtrac, S.A.P.I. De C.V., 45601 Tlaquepaque, J...",3,2026-02-08 17:19:56,2026-02-09 00:33:33,190-08B26AA,"Agtrac, S.A.P.I. De C.V.",Desconocido
4,NaN,FR1706791,CASE IH,TRACTORS,UTILITY,FARMALL® JXM,Farmall® 90JXM,696055310,A.A.02.013.005.004,1602-38000BFA,...,CS,TRACTORS,CM90,Default Company,23,2026-02-08 11:18:29,2026-02-09 00:33:33,Desconocido,Desconocido,Desconocido


**Fuentes de datos:** La información proviene del archivo PopulationView_2026.xlsx, el cual contiene el registro de equipos con telemetría asociados a la empresa.
En esta base se incluye información sobre identificación de equipos, características técnicas, ubicación geográfica, distribuidores y datos operativos provenientes del sistema de telemetría.

**Obtención de la información:** Los datos provienen de los sistemas que monitorean a los equipos mediante dispositivos de telemetría instalados en cada unidad. Esta información se registra automáticamente y posteriormente se exporta a un archivo de Excel.




A continuación un diccionario sobre el contenido que se encuentra en el excel:

**Diccionario de datos**

1. **VIN:** Identificador único del vehículo o equipo. Puede aparecer en diferentes formatos, como de 9 o 17 caracteres o la versión estándar utilizada para identificar cada unidad.

2. **Brand:** Habla sobre la marca del equipo.

3. **Type:** Describe el tipo de maquinaria o vehículo.

4. **Product:** Indica la categoría del producto al que pertenece el equipo.

5. **Series:** Identificador interno del modelo dentro de una línea de producción.

6. **Model:** Mdelo específico del equipo.

7. **Technical type:** Clasifica el tipo específico de equipo.

8. **Ice code:** Código que identifica el tipo de motor del equipo.

9. **Device serial number**: Número de serie del dispositivo telemático que se encuentra instalado en el equipo.

10. **Severity level:** Nivel de severidad (grado de intensidad, gravedad o impacto) de las alertas registradas por el sistema.

11. **Context:** Contexto o tipo de evento registrado en el sistema.

12. **Dealer:** Distribuidor asociado al equipo.

13. **Dealer name:** Es el nombre del distribuidor responsable del equipo.

14. **Dealer code:** Código que identifica al distribuidor dentro del sistema.

15. **Country of competence:**  País donde opera el equipo o donde el distribuidor tiene competencia.

16. **Latitude:** Latitud de la última ubicación registrada del equipo.

17. **Longitude:** Longitud de la última ubicación registrada del equipo.

18. **Last known position:** País donde se registró por última vez la ubicación del equipo.

19. **Telematics box:** Tipo de dispositivo de telemetría instalado en el equipo.

20. **Telematics brand:** Marca del sistema de telemetría utilizado.

21. **Telematics type:** Describe el tipo de tecnología telemática utilizada por el equipo.

22. **Telematics model:** Modelo específico del dispositivo telemático instalado.

23. **Owner name:** Nombre del propietario del equipo.

24. **Engine hours:** Total de horas acumuladas de operación del motor.

25. **Latest communication date:** Es la fecha en la que el equipo se comunicó por última vez con el sistema de telemetría.

26. **BDE processed at time:** Fecha y hora en que la información fue procesada dentro del sistema.

27. **Lower pin break y upper pin break:** Son indicadores técnicos relacionados con algunnos parámetros del sistema del equipo.

28. **Myfleet dealer:** Distribuidor dentro del sistema MyFleet.

29. **Myfleet dealer name:** Nombre del distribuidor registrado en el sistema MyFleet.

30. **EULA accepted y EULA date:** Indica si el usuario aceptó los términos y en que fecha se aceptaron los términos de uso del sistema.

# **Horas 2024-2025.xlsx**
El archivo Horas permite observar las horas de funcionamiento de los equipos a lo largo del tiempo, mientras que Reporte contiene información operativa diaria de las unidades. Por esta razón, **ambos archivos se utilizan para analizar la actividad y comportamiento operativo de los equipos.**

In [6]:
print("Horas 2024-2025.xlsx")
df_horas.head()

Horas 2024-2025.xlsx


,Num,Distribuidor,IMEI,Alias,Marca,Modelo,Ciudad,Latitud,Longitud,2024-01,...,2024-03,2024-04,2024-05,2024-06,2024-07,2024-08,2024-09,2024-10,2024-11,2024-12
0,1,AGROEQUIPOS,353738101215364,353738101215364,NEW HOLLAND CE,B80C,Cintalapa,16.690481,-93.747589,0,...,0,0,0,0,0,0,0,0,0,0
1,2,AGROEQUIPOS,353738101216677,353738101216677,NEW HOLLAND AG,7610S HERENCIA,Chahuites,16.281037,-94.198885,4,...,0,0,10,0,23,14,10,1,45,0
2,3,AGROEQUIPOS,353738101232583,353738101232583,NEW HOLLAND AG,6810,Chahuites,16.309873,-94.187625,0,...,0,0,0,0,0,0,76,87,108,67
3,4,AGROEQUIPOS,353738101280137,353738101280137,NEW HOLLAND AG,7610S HERENCIA,0,16.302366,-94.190832,79,...,23,27,24,10,2,59,8,48,48,29
4,5,AGROEQUIPOS,353738102813829,353738102813829,NEW HOLLAND AG,6810,0,16.281820,-94.198922,43,...,26,0,0,0,0,0,0,0,0,0


**Fuentes de datos**
1. **Horas:** Es una base con información mensual de actividad de los equipos monitoreados. Cada fila representa un equipo y las columnas mensuales permiten observar su comportamiento a lo largo del tiempo.

**¿Cómo se obtiene la información?**

La información se obtiene a partir de los sistemas que monitorean los equipos mediante dispositivos de telemetría instalados en cada unidad.
Por lo que, estos sistemas registran automáticamente datos relacionados con su funcionamiento, actividad y operación. Después, la información se exporta a archivos de Excel para su consulta, análisis y seguimiento.





**Diccionarios - Horas**

1. **Distribuidor:** Es el distribuidor asociado al equipo.

2. **IMEI:** Identificador del dispositivo de comunicación instalado en el equipo.

3. **Alias:** Nombre con el que se reconoce al equipo dentro del sistema.

4. **Marca:** Marca del equipo.

5. **Modelo:** Indica el modelo específico del equipo.

6. **Ciudad:** Muestra la ciudad asociada al equipo.

7. **Estado:**: Muestra el estado o región donde se ubica el equipo.

8. **Latitud:** Indica la latitud de la ubicación registrada del equipo.

9. **Longitud:** Indica la longitud de la ubicación registrada del equipo.

10. **Columnas mensuales (por ejemplo 2024-01, 2024-02, 2024-03...):**
Representan los valores de actividad registrados por equipo en cada mes, lo que permite analizar su comportamiento a lo largo del tiempo.

# **Reporte_unidades_dia_anterior.xlsx**

In [7]:
print("Reporte_unidades_dia_anterior.xlsx")
df_unidades.head()

Reporte_unidades_dia_anterior.xlsx


,16397000062879,16397000062879.1,CASE CE,Desconocido,Desconocido.1,Unnamed: 5,Desconocido.2,Desconocido.3,Desconocido.4,Desconocido.5,...,-100.365349,0.6,0.7,0.8,0.9,0.10,0.11,0.12,9,Unnamed: 35
0,352556102049694,352556102049694,NEW HOLLAND AG,TS6.120,FWD 8X8 PS,2022-06-02,Desconocido,Desconocido,Desconocido,Desconocido,...,0.000000,0.0,0.0,0,0.0,0.0,0.0,0,9,4.0
1,352556102529786,352556102529786,NEW HOLLAND AG,Desconocido,Desconocido,2024-02-09,ZEBG01193,TRACSAGRAL,San Pedro Tlaquepaque,Jalisco,...,-103.342296,0.0,0.0,0,0.0,0.0,24.0,0,9,2.0
2,352556102532012,352556102532012,Desconocido,Desconocido,Desconocido,NaT,serieprueba,TRACSAGRAL,Tala,Jalisco,...,-103.660210,0.0,0.0,0,0.0,0.0,24.0,0,9,NaN
3,352556102805988,352556102805988,Desconocido,Desconocido,Desconocido,NaT,Desconocido,Desconocido,Desconocido,Desconocido,...,-98.960082,0.0,0.0,0,0.0,0.0,0.0,0,9,NaN
4,352556102809907,ATN - 352556102809907,Desconocido,Desconocido,Desconocido,NaT,Desconocido,ATN,Altamira,Tamaulipas,...,-97.938348,0.0,0.0,0,0.0,0.0,24.0,0,9,NaN


**Fuentes de datos**
1. **Unidades:** Es una base con información operativa diaria de las unidades. Incluye datos relacionados con la operación, uso, ubicación y desempeño reciente de los equipos.


**Diccionario - Reporte unidades**

1. **No:** Es un número consecutivo que identifica cada registro dentro del reporte.

2. **IMEI:** Identifica el dispositivo telemático instalado en el equipo.

3. **Alias**: Nombre del equipo dentro del sistema.

4. **Marca:** Marca del equipo registrada en el sistema.

5. **Modelo:** Describe el modelo específico del equipo.

6. **Configuración:** Configuración o variante del equipo.

7. **Fecha alta:** Representa la fecha en que el equipo fue registrado.

8. **No. serie:**
Es el número de serie del equipo.

9. **Distribuidor:**
Distribuidor asociado al equipo.

10. **Ciudad:**
Ciudad vinculada al equipo.

11. **Estado:**
Estado o región donde se ubica el equipo.

12. **Horómetro:** Representa el total de horas acumuladas de operación del equipo.

13. **Odómetro:** Distancia total recorrida por el equipo.

14. **Total:** Valor total calculado dentro del reporte.

15. **Cerrados:** Número de servicios  que han sido cerrados.

16. **C. fuera:** Cantidad de servicios cerrados fuera del proceso habitual.

17. **Pendientes:**
Representa la cantidad de servicios que aún están pendientes.

18. **Indicadores de mantenimiento:**
Intervalos de mantenimiento asociados al equipo.

19. **Cliente:**
Es el nombre del cliente asociado al equipo.

20. **Teléfono:**
Es el número telefónico del cliente.

21. **Email:**
Es el correo electrónico del cliente.

22. **Latitud:**
Latitud de la ubicación registrada del equipo.

23. **Longitud:**
Longitud de la ubicación registrada del equipo.

24. **PEP:**
Código interno utilizado dentro del sistema.

25. **Combustible:**
Información relacionada con el consumo de combustible.

26. **Distancia:**
Distancia recorrida por el equipo.

27. **Vel. máxima (kph):**
Velocidad máxima registrada para el equipo.

28. **Vel. promedio (kph)**
Velocidad promedio registrada.

29. **T. encendido (hrs)**
Tiempo total que el equipo estuvo encendido.

30. **T. apagado (hrs):**
Tiempo total que el equipo estuvo apagado

# **Indicadores de Ruteo.xlsx**
Este archivo contiene información relacionada con los indicadores de ruteo de los equipos, lo que permite analizar cómo se están utilizando las unidades durante su operación.

**¿Cómo se obtiene la información?**

Los datos se generan a partir de los sistemas de gestión logística y monitoreo de operaciones, los cuales registran información sobre las rutas realizadas por las unidades, distancias recorridas y cumplimiento de actividades programadas.

In [8]:
print("Indicadores de Ruteo.xlsx")
df_rutas.head()

Indicadores de Ruteo.xlsx


,Division,Distancia del Dia,Distancia de Ruta
0,NaN,4381.0,4381.0
1,ADMIN.VRM,69138.9,69138.9
2,AGROCISAGRAL,1895107.2,1895107.2
3,AGROEQUIPOS,589540.9,589540.9
4,AGROINDUSTRIALES,7235.8,7235.8


**Fuente de datos**
Contiene información relacionada con el desempeño operativo de las rutas utilizadas por las diferentes divisiones o distribuidores. Este excel permite analizar indicadores de operación como distancia recorrida, cumplimiento de horarios, nivel de utilización y visitas realizadas.

La información proviene de los sistemas de monitoreo operativo y de gestión de rutas de la empresa, donde se registran las actividades realizadas por las unidades durante sus recorridos.

**Diccionario de datos**

1. **Division:** Indica la división, área o distribuidor responsable de la operación registrada.

2. **Distancia del día:** Representa la distancia total recorrida durante el día por las unidades asociadas a esa división.

3. **Distancia de ruta:**
Indica cuál es la distancia total correspondiente a las rutas asignadas para la operación.

4. **Cumplimiento salida:**
Mide el nivel de cumplimiento respecto al horario establecido de salida de las rutas.

5. **%Utilización:** Representa el porcentaje de los recursos o unidades durante la operación.

6. **% Visitas plan:**
Indica el porcentaje de visitas planificadas que fueron realizadas.

7. **%Visitas dlientes:** Mide el porcentaje de visitas efectivas realizadas a clientes.

8. **Total kilos:** Representa el total de kilos transportados o movilizados durante la operación.

# **Mantenimientos 2024-2025.xlsx**

**¿Cómo se obtiene la información?**

La información proviene de los sistemas internos de gestión de mantenimiento de la empresa, donde se registran las actividades de servicio realizadas a los equipos.

In [9]:
print("Mantenimientos 2024-2025.xlsx")
df_mantenimiento.head()

Mantenimientos 2024-2025.xlsx


,ALIAS,MARCA,MODELO,CONFIGURACION,NO SERIE,DISTRIBUIDOR,SERVICIO,HRMTRO,ACTUAL,TIPO,FECHA,ESTATUS
0,DASSA - NH09655M,NEW HOLLAND AG,TS6.125,FWD 16X8 PS,NH09655M-001854359,DASSA,3900,3900.0,6418,Preventivo,2024-05-01 10:23:00,Pendiente
1,ARBSA - S508144M,NEW HOLLAND AG,7610S HERENCIA,4WD,S508144M-8105974,ARBSA,1500,2055.0,2706,Preventivo,2024-05-01 10:23:00,Cerrada
2,DASSA - S508370M,NEW HOLLAND AG,7610S HERENCIA,4WD,S508370M-8107452,DASSA,300,305.0,1452,Preventivo,2024-05-01 10:23:00,Cerrada
3,MADISA-S204600M,NEW HOLLAND AG,6610S HERENCIA,2WD,S204600M-8087448,MADISA,3300,3302.0,6271,Preventivo,2024-05-01 10:23:00,Pendiente
4,ARBSA - S306326M,NEW HOLLAND AG,6610S HERENCIA,4WD,S306326M-8088579,ARBSA,900,901.0,2355,Preventivo,2024-05-01 10:23:00,CerradaFuera


**Fuentes de datos**
El archivo contiene información relacionada con los servicios y mantenimientos realizados a los equipos durante el periodo 2024-2025. Estos datos permiten analizar el estado de las unidades y dar seguimiento a las actividades de mantenimiento realizadas.

**Diccionarios**

1. **No:** Es el número de registro del mantenimiento dentro del sistema.

2. **Alias:** Nombre del equipo dentro de la operación.

3. **Marca:**
Marca del equipo o maquinaria.

4. **Modelo:** Modelo específico del equipo.

5. **Configuración:**
Configuración técnica del equipo o versión del modelo.

6. **No serie:**
Número de serie único del equipo.

7. **Distribuidor:**
Responsable del equipo o del servicio realizado.

8. **Servicio:**
Tipo de servicio realizado al equipo.

9. **HRMTRO:** Horas registradas en el horómetro del equipo al momento del servicio.

10. **Actual:** Indica si el registro corresponde al estado actual del mantenimiento.

11. **Tipo:**
Clasificación del mantenimiento realizado.

12. **Fecha:** Que día se realizó o registró el servicio de mantenimiento.

13. **O.T.:** Número de orden de trabajo asociada al mantenimiento.

14. **Estatus:**
Estado del servicio registrado.

15. **Editar:** Campo utilizado en el sistema para edición o modificación del registro.

# **Distribuidores_Todo.xlsx**

**¿Cómo se obtiene la información?**

La información se obtiene mediante:
* Los equipos registran automáticamente sus horas de operación mediante sistemas de telemetría o registros operativos.
* Los datos recopilados se almacenan en bases de datos de monitoreo y gestión de maquinaria.
* Se analizan las horas acumuladas de cada equipo y se ordenan de mayor a menor.
* Para cada distribuidor se seleccionan los 50 equipos con mayor número de horas de operación.


In [10]:
print("Distribuidores_Todo.xlsx")
df_top50.head()

Distribuidores_Todo.xlsx


,No.,Unidad,Distribuidor,Marca,Modelo,Estado,Ciudad,Horas
0,1,IAFSA - JJAM0280,INSUMOS AGROPECUARIOS FRONTERIZOS SA DE CV,CASE IH,Magnum 290,Tamaulipas,Valle Hermoso,1216.46
1,2,IAFSA - 4100035M,INSUMOS AGROPECUARIOS FRONTERIZOS SA DE CV,NEW HOLLAND AG,TS6.120 ELITE,Tamaulipas,Valle Hermoso,1104.18
2,3,IAFSA - NKJ0B80CTNKH30831,INSUMOS AGROPECUARIOS FRONTERIZOS SA DE CV,NEW HOLLAND CE,B80C,Nuevo León,General Escobedo,478.42
3,4,IAFSA - NKJ0B80CTNKH30828,INSUMOS AGROPECUARIOS FRONTERIZOS SA DE CV,NEW HOLLAND CE,B80C,Nuevo León,Galeana,454.09
4,5,IAFSA - NKJ0B80CKNKH30816,INSUMOS AGROPECUARIOS FRONTERIZOS SA DE CV,NEW HOLLAND CE,B80C,Nuevo León,Guadalupe,420.73


**Fuente de datos**
Los datos del Top 50 provienen de los sistemas de monitoreo de equipos utilizados por la empresa para registrar la operación de maquinaria en campo.
Cada archivo Excel corresponde al Top 50 de equipos con mayor número de horas de operación para un distribuidor específico.

**Diccionario de datos: Top 50**

Los Top 50 de cada distribuidor normalmente tienen las mismas columnas, por eso solo necesitas revisar un archivo para hacer el diccionario.

1. **No.:**
Número consecutivo que indica la posición del equipo dentro del ranking del Top 50.

2. **Unidad:**
Identificador del equipo dentro del sistema. Generalmente combina el nombre del distribuidor con el número de serie o código interno del equipo.

3. **Distribuidor:**
Empresa o concesionario responsable del equipo y encargado de su seguimiento operativo.

4. **Marca:**
Marca del equipo o fabricante de la maquinaria.

5. **Modelo:**
Modelo específico del equipo registrado.

6.**Estado:**
Estado de la República donde se encuentra operando el equipo.

7.**Ciudad:**
Ciudad o localidad donde se ubica el equipo.

8.**Horas:**
Total de horas acumuladas de operación del equipo registradas por el sistema. Esta variable es la que determina el ranking dentro del Top 50.

## 2. Dimensiones de los DataFrames (Filas y Columnas)

In [11]:
print("→ df_population Shape")
print(df_population.shape)
print("\n→ df_horas Shape")
print(df_horas.shape)
print("\n→ df_unidades Shape")
print(df_unidades.shape)
print("\n→ df_rutas Shape")
print(df_rutas.shape)
print("\n→ df_mantenimiento Shape")
print(df_mantenimiento.shape)
print("\n→ df_top50 Shape")
print(df_top50.shape)

→ df_population Shape
(535, 30)

→ df_horas Shape
(9405, 21)

→ df_unidades Shape
(10845, 36)

→ df_rutas Shape
(35, 3)

→ df_mantenimiento Shape
(23869, 12)

→ df_top50 Shape
(29, 8)


**df_population (585 registros, 34 columnas):** Es un dataset de tamaño moderado, funcionando como la base principal o inventario.

**df_horas (9405 registros, 22 columnas):** Contiene un número considerable de registros de actividad mensual, lo que sugiere una precisión detallada en el tiempo para el monitoreo de uso de las unidades.

**df_unidades (10846 registros, 38 columnas):** Es el dataset con más filas, lo que indica que contiene información operativa diaria muy detallada.

**df_rutas (35 registros, 8 columnas):** Es un dataset pequeño con datos agregados o de resumen sobre indicadores de ruteo.

**df_mantenimiento (23904 registros, 15 columnas):** Es el dataset con más registros, lo que proporciona una visión detallada de los servicios de mantenimiento de las unidades.

**df_top50 (29 registros, 8 columnas):** Es un dataset muy pequeño que contiene información muy específica y resumida (como un ranking).

## 3. Nombres de las Columnas

In [12]:
print("→  df_population Columns")
print(df_population.columns)
print("\n→  df_horas Columns")
print(df_horas.columns)
print("\n→  df_unidades Columns")
print(df_unidades.columns)
print("\n→  df_rutas Columns")
print(df_rutas.columns)
print("\n→  df_mantenimiento Columns")
print(df_mantenimiento.columns)
print("\n→  df_top50 Columns")
print(df_top50.columns)

→  df_population Columns
Index(['VIN (17 CHARACTERS)', 'VIN', 'BRAND', 'TYPE', 'PRODUCT', 'SERIES',
       'MODEL', 'TECHNICAL TYPE', 'ICE CODE', 'DEVICE SERIAL NUMBER',
       'SEVERITY LEVEL', 'CONTEXT', 'DEALER', 'DEALER NAME', 'DEALER CODE',
       'COUNTRY OF COMPETENCE', 'LATITUDE', 'LONGITUDE',
       'LAST KNOWN POSITION (COUNTRY)', 'TELEMATICS BOX', 'TELEMATICS BRAND',
       'TELEMATICS TYPE', 'TELEMATICS MODEL', 'OWNER NAME', 'ENGINE HOURS',
       'LATEST COMMUNICATION DATE', 'BDE PROCESSED AT TIME', 'MYFLEET DEALER',
       'MYFLEET DEALER NAME', 'EULA ACCEPTED'],
      dtype='str')

→  df_horas Columns
Index(['Num', 'Distribuidor', 'IMEI', 'Alias', 'Marca', 'Modelo', 'Ciudad',
       'Latitud', 'Longitud', '2024-01', '2024-02', '2024-03', '2024-04',
       '2024-05', '2024-06', '2024-07', '2024-08', '2024-09', '2024-10',
       '2024-11', '2024-12'],
      dtype='str')

→  df_unidades Columns
Index([    16397000062879, '16397000062879.1',          'CASE CE',
            '

Podemos observar que es necesario realizar una estandarización en los nombres de las columnas de los DataFrames. Algunos usan mayúsculas (`BRAND`, `TYPE`), otros usan CamelCase (`Fecha Alta`, `No. Serie`), y algunos incluyen caracteres especiales o espacios (`VIN (9 CHARACTERS)`, `Vel. Maxima (kph)`, `Distancia del Dia`).

Existen columnas que representan la misma información pero tienen nombres diferentes en distintos DataFrames, lo que requerirá que renombremos esas columnas para facilitar las uniones. **Ejemplos:**

  * `BRAND` (df_population) vs `Marca` (df_horas, df_unidades, df_top50) vs `MARCA` (df_mantenimiento)
  *   `MODEL` (df_population) vs `Modelo` (df_horas, df_unidades, df_top50) vs `MODELO` (df_mantenimiento)
  *   `DEALER` (df_population) vs `Distribuidor` (df_horas, df_unidades, df_top50) vs `DISTRIBUIDOR` (df_mantenimiento) vs `Division` (df_rutas)
  *   `VIN` (df_population) vs `No. Serie` (df_unidades) vs `NO SERIE` (df_mantenimiento)
  *   `IMEI` (df_horas, df_unidades) y `DEVICE SERIAL NUMBER` (df_population) aparentemente se refieren al mismo identificador.


Hay columnas con información repetida, como `df_population` que tiene `VIN (9 CHARACTERS)`, `VIN (17 CHARACTERS)`, y `VIN`, que parecen ser diferentes representaciones del mismo identificador. Esto podría simplificarse a una sola columna.

## 4. Tipos de Datos

In [13]:
print("df_population dtypes")
df_population.dtypes

df_population dtypes


VIN (17 CHARACTERS)                         str
VIN                                         str
BRAND                                       str
TYPE                                        str
PRODUCT                                     str
SERIES                                      str
MODEL                                       str
TECHNICAL TYPE                            int64
ICE CODE                                    str
DEVICE SERIAL NUMBER                        str
SEVERITY LEVEL                              str
CONTEXT                                     str
DEALER                                      str
DEALER NAME                                 str
DEALER CODE                                 str
COUNTRY OF COMPETENCE                       str
LATITUDE                                float64
LONGITUDE                               float64
LAST KNOWN POSITION (COUNTRY)               str
TELEMATICS BOX                              str
TELEMATICS BRAND                        

###**df_population:**

  *   `VIN (9 CHARACTERS)`, `VIN (17 CHARACTERS)`, `VIN` son de tipo `object`, lo cual es adecuado ya que son identificadores alfanuméricos.

  *   `BRAND`, `TYPE`, `PRODUCT`, `MODEL`, `ICE CODE`, `DEVICE SERIAL NUMBER`, `SEVERITY LEVEL`, `CONTEXT`, `DEALER`, `DEALER NAME`, `DEALER CODE`, `COUNTRY OF COMPETENCE`, `LAST KNOWN POSITION (COUNTRY)`, `TELEMATICS BOX`, `TELEMATICS BRAND`, `TELEMATICS TYPE`, `TELEMATICS MODEL`, `OWNER NAME`, `LOWER PIN BREAK`, `UPPER PIN BREAK`, `MYFLEET DEALER`, `MYFLEET DEALER NAME`, `EULA ACCEPTED` son `object`, lo que es apropiado para datos categóricos o de texto.
  
  *   `TECHNICAL TYPE` y `ENGINE HOURS` son `int64`, lo cual es correcto para valores numéricos enteros.

  *   `LATITUDE` y `LONGITUDE` son `float64`, adecuado para coordenadas geográficas.

  *   `LATEST COMMUNICATION DATE`, `BDE PROCESSED AT TIME`, `EULA DATE` son `datetime64[ns]`, lo que indica que Pandas ha reconocido correctamente estos campos como fechas y horas, lo cual es óptimo para el análisis.

In [14]:
print("\ndf_horas dtypes")
df_horas.dtypes


df_horas dtypes


Num               int64
Distribuidor        str
IMEI              int64
Alias            object
Marca            object
Modelo           object
Ciudad           object
Latitud         float64
Longitud        float64
2024-01           int64
2024-02           int64
2024-03           int64
2024-04           int64
2024-05           int64
2024-06           int64
2024-07           int64
2024-08           int64
2024-09           int64
2024-10           int64
2024-11           int64
2024-12           int64
dtype: object

###**df_horas:**
  *   `Num`, `IMEI`, y las columnas mensuales (`2024-01` a `2024-12`) son `int64`, que es el tipo correcto para conteos o identificadores numéricos. Las columnas mensuales parecen representar horas de uso, por lo que `int64` es adecuado.
  
  *   `Distribuidor`, `Alias`, `Marca`, `Modelo`, `Ciudad`, `Estado` son `object`, lo que es esperado para texto o categorías.
  
  *   `Latitud` y `Longitud` son `float64`, correcto para los datos geográficos.


In [15]:
print("\ndf_unidades dtypes")
df_unidades.dtypes


df_unidades dtypes


 16397000062879              object
 16397000062879.1            object
 CASE CE                        str
 Desconocido                 object
 Desconocido.1                  str
 Unnamed: 5          datetime64[us]
 Desconocido.2               object
 Desconocido.3                  str
 Desconocido.4                  str
 Desconocido.5                  str
 0                          float64
 0.1                        float64
 0.2                          int64
 0.3                          int64
 0.4                          int64
 0.5                          int64
 Unnamed: 16                    str
 Unnamed: 17                    str
 Unnamed: 18                    str
 Unnamed: 19                    str
 Unnamed: 20                    str
 Unnamed: 21                    str
 Unnamed: 22                    str
 Unnamed: 23                    str
 Unnamed: 24                    str
 25.709133                  float64
-100.365349                 float64
 0.6                        

###**df_unidades:**
    
  *   `No`, `Total`, `Cerrados`, `C.Fuera`, `Pendientes`, `Vel. Maxima (kph)` son `int64`, lo cual es adecuado para conteos o valores discretos.
  
  *   `Horometro`, `Odometro`, `Latitud`, `Longitud`, `Combustible`, `Distancia`, `Vel. Promedio (kph)`, `T. Encendido (hrs)`, `T. Apagado (hrs)` son `float64`, apropiado para valores numéricos con decimales.
  
  *   `IMEI`, `Alias`, `Marca`, `Modelo`, `Configuracion`, `No. Serie`, `Distribuidor`, `Ciudad`, `Estado`, `50`, `300`, `600`, `900`, `1200`, `1500`, `1800`, `2100`, `2400`, `Cliente`, `Telefono`, `Email` son `object`. Las columnas numéricas (`50` a `2400`) siendo `object` son una inconsistencia importante y probablemente deberían ser numéricas. `Fecha Alta` es `datetime64[ns]`, lo cual es correcto.

In [16]:
print("\ndf_rutas dtypes")
df_rutas.dtypes


df_rutas dtypes


Division                 str
Distancia del Dia    float64
Distancia de Ruta    float64
dtype: object

###**df_rutas:**
   
  *   `Distancia del Dia`, `Distancia de Ruta`, `Total Kilos` son `float64`, lo cual es correcto para métricas continuas.
  
  *   `Cumplimiento Salida (06:30:00)`, `% Utilizacion`, `% Visitas Plan`, `% Visitas Clientes` son `int64`, adecuado para porcentajes o conteos.
  
  *   `Division` es `object`, lo que es esperado para categorías de texto.

In [17]:
print("\ndf_mantenimiento dtypes")
df_mantenimiento.dtypes


df_mantenimiento dtypes


ALIAS             object
MARCA                str
MODELO            object
CONFIGURACION        str
NO SERIE          object
DISTRIBUIDOR         str
SERVICIO           int64
HRMTRO           float64
ACTUAL             int64
TIPO                 str
FECHA             object
ESTATUS              str
dtype: object

###**df_mantenimiento:**
    
  *   `SERVICIO` y `ACTUAL` son `int64`, lo cual es apropiado.
  
  *   `HRMTRO` es `float64`, adecuado para horas con decimales.
  
  *   `ALIAS`, `MARCA`, `MODELO`, `CONFIGURACION`, `NO SERIE`, `DISTRIBUIDOR`, `TIPO`, `O.T.`, `ESTATUS` son `object`, esperado para texto y categorías.
  
  *   `FECHA` es `object` y deberíamos convertirlo a `datetime` para permitir un análisis correcto.

In [18]:
print("\ndf_top50 dtypes")
df_top50.dtypes


df_top50 dtypes


No.               int64
Unidad              str
Distribuidor        str
Marca               str
Modelo              str
Estado              str
Ciudad              str
Horas           float64
dtype: object

###**df_top50:**
    
  * `No.`, `Horas` son `int64` y `float64`, lo cual es correcto para un ranking y horas acumuladas.
  
  *   `Unidad`, `Distribuidor`, `Marca`, `Modelo`, `Estado`, `Ciudad` son `object`, lo que es adecuado para texto o categorías.

## 5. Información General y Conteo de Valores No-Nulos

In [19]:
print("→  df_population Info")
df_population.info()
print("\n→  df_horas Info")
df_horas.info()
print("\n→  df_unidades Info")
df_unidades.info()
print("\n→  df_rutas Info")
df_rutas.info()
print("\n→  df_mantenimiento Info")
df_mantenimiento.info()
print("\n→  df_top50 Info")
df_top50.info()

→  df_population Info
<class 'pandas.DataFrame'>
RangeIndex: 535 entries, 0 to 534
Data columns (total 30 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   VIN (17 CHARACTERS)            242 non-null    str           
 1   VIN                            535 non-null    str           
 2   BRAND                          535 non-null    str           
 3   TYPE                           535 non-null    str           
 4   PRODUCT                        535 non-null    str           
 5   SERIES                         535 non-null    str           
 6   MODEL                          535 non-null    str           
 7   TECHNICAL TYPE                 535 non-null    int64         
 8   ICE CODE                       535 non-null    str           
 9   DEVICE SERIAL NUMBER           535 non-null    str           
 10  SEVERITY LEVEL                 535 non-null    str           
 11  CONTEXT 

*   **df_population:** Presenta varias columnas con un número considerable de valores nulos. Estas columnas necesitan un tipo de tratamiento debido a la cantidad de información faltante.

*   **df_horas:** Muestra una cantidad significativa de nulos en `Modelo` y `Estado` (ambos con 3142 no nulos de 9405), lo que sugiere que aproximadamente dos tercios de los registros carecen de esta información. `Marca`, `Ciudad`, `Latitud` y `Longitud` tienen muy pocos nulos (9390 de 9405 no nulos).

*   **df_unidades:** Muchas columnas presentan valores nulos. Las columnas `Marca`, `Modelo`, `Configuracion`, `Fecha Alta`, `No. Serie`, `Distribuidor`, `Ciudad`, `Estado`, `Cliente`, `Telefono`, `Email` tienen un porcentaje variable de nulos, siendo algunas bastante elevadas. Además, las columnas `50` a `2400` tienen un número muy bajo de valores no nulos (124 de 10846), lo que indica que están casi vacías y probablemente no son útiles, se tendrían que revisar a detalle.

*   **df_rutas:** `Total Kilos` no tiene ningún valor no nulo (0 de 35), lo que significa que esta columna está completamente vacía y no aporta información. La columna `Division` tiene 34 valores no nulos de 35, con un solo valor nulo, lo cual es fácilmente manejable.

*   **df_mantenimiento:** Las columnas `No` y `Editar` no tienen ningún valor no nulo (0 de 23904), indicando que están completamente vacías y pueden ser eliminadas. `O.T.` (Orden de Trabajo) tiene un alto número de nulos (9484 de 23904 no nulos), lo que sugiere que muchas órdenes de trabajo no están registradas. Otras columnas como `MARCA`, `MODELO`, `CONFIGURACION`, `NO SERIE`, `DISTRIBUIDOR`, `TIPO`, `ESTATUS` tienen muy pocos nulos, lo que las hace confiables.

*   **df_top50:** Todas las columnas tienen 29 valores no nulos, lo que significa que este DataFrame está completo y no presenta valores faltantes.

## 6. Análisis de Valores Nulos

In [20]:
print("Valores Nulos en df_population")
nulos_population = df_population.isnull().sum()
porcentaje_nulos_population = (nulos_population / len(df_population)) * 100
pd.DataFrame({"Nulos": nulos_population, "Porcentaje": porcentaje_nulos_population}).sort_values(by="Porcentaje", ascending=False)

Valores Nulos en df_population


,Nulos,Porcentaje
VIN (17 CHARACTERS),293,54.766355
VIN,0,0.000000
BRAND,0,0.000000
TYPE,0,0.000000
PRODUCT,0,0.000000
SERIES,0,0.000000
MODEL,0,0.000000
TECHNICAL TYPE,0,0.000000
ICE CODE,0,0.000000
DEVICE SERIAL NUMBER,0,0.000000


**df_population:**
    
  *   `VIN (17 CHARACTERS)`: 53.33% de valores nulos. Es una cantidad considerable que sugiere un posible problema de registro o que esta versión del VIN no siempre está disponible.
  
  *   `DEALER NAME`: 72.48% de valores nulos. Indica que la información del nombre del distribuidor está mayormente ausente.
  
  *   `DEALER CODE`: 48.46% de valores nulos. Cerca de la mitad de los registros carecen de este código.
  
  *   `LOWER PIN BREAK`: 63.76% de valores nulos. Una gran proporción de datos faltantes para este indicador.
  
  *   `UPPER PIN BREAK`: 91.31% de valores nulos. Casi todos los registros carecen de esta información, lo que parece demostrar que esta columna tiene una baja utilidad.
  
  *   `MYFLEET DEALER` y `MYFLEET DEALER NAME`: Ambos con 46.50% de nulos. Es importante considerar si esta información es crítica y cómo imputarla si es necesario.
  
  *   `EULA ACCEPTED`: 67.69% de nulos. La mayoría de los registros no indican si se aceptaron los términos de uso.
  
  *   `EULA DATE`: 92.14% de nulos. Similar a `EULA ACCEPTED`, esta columna está casi vacía.

In [21]:
print("\nValores Nulos en df_horas")
nulos_horas = df_horas.isnull().sum()
porcentaje_nulos_horas = (nulos_horas / len(df_horas)) * 100
pd.DataFrame({"Nulos": nulos_horas, "Porcentaje": porcentaje_nulos_horas}).sort_values(by="Porcentaje", ascending=False)


Valores Nulos en df_horas


,Nulos,Porcentaje
Modelo,6263,66.592238
Distribuidor,0,0.000000
IMEI,0,0.000000
Alias,0,0.000000
Num,0,0.000000
Marca,0,0.000000
Ciudad,0,0.000000
Latitud,0,0.000000
Longitud,0,0.000000
2024-01,0,0.000000


**df_horas:**

  *   `Modelo` y `Estado`: Ambos con 66.59% de valores nulos. Esto es un porcentaje muy alto y es un problema para cualquier análisis que dependa de estas columnas.
  
  *   `Marca`, `Ciudad`, `Latitud`, `Longitud`: Presentan un porcentaje muy bajo de nulos (0.16%), lo cual es manejable mediante imputación.

In [22]:
print("\nValores Nulos en df_unidades")
nulos_unidades = df_unidades.isnull().sum()
porcentaje_nulos_unidades = (nulos_unidades / len(df_unidades)) * 100
pd.DataFrame({"Nulos": nulos_unidades, "Porcentaje": porcentaje_nulos_unidades}).sort_values(by="Porcentaje", ascending=False)


Valores Nulos en df_unidades


,Nulos,Porcentaje
Unnamed: 24,9643,88.916551
Unnamed: 23,9252,85.311203
Unnamed: 22,8774,80.903642
Unnamed: 21,8195,75.564776
Unnamed: 20,7537,69.497464
Unnamed: 19,6742,62.166897
Unnamed: 18,5904,54.439834
Unnamed: 17,4931,45.467958
Unnamed: 16,4015,37.021669
Unnamed: 5,641,5.910558


**df_unidades:**

  *   Las columnas `50` a `2400`: Tienen un porcentaje demaciado alto de valores nulos (98.85% - 99.63%). Esto las hace prácticamente inútiles para un buen análisis.
  
  *   `Cliente`, `Telefono`, `Email`: Todas superan el 97% de nulos, indicando que la información de contacto del cliente está casi ausente y fueron eliminadas.
  
  *   `PEP`: 100% de nulos. Columna completamente vacía, es muy probable que sea eliminada.
  
  *   `Marca`, `No. Serie`: Tienen alrededor del 6.5% de nulos. Estas columnas con un porcentaje moderado de nulos pueden ser imputadas.

In [23]:
print("\nValores Nulos en df_rutas")
nulos_rutas = df_rutas.isnull().sum()
porcentaje_nulos_rutas = (nulos_rutas / len(df_rutas)) * 100
pd.DataFrame({"Nulos": nulos_rutas, "Porcentaje": porcentaje_nulos_rutas}).sort_values(by="Porcentaje", ascending=False)


Valores Nulos en df_rutas


,Nulos,Porcentaje
Division,1,2.857143
Distancia del Dia,0,0.000000
Distancia de Ruta,0,0.000000


**df_rutas:**
  
  *   `Total Kilos`: Tiene el 100% de valores nulos. Esta columna está completamente vacía y seguramente sea eliminada.
  
  *   `Division`: 2.86% de valores nulos. Un porcentaje bajo que puede imputarse con la moda.

In [24]:
print("\nValores Nulos en df_mantenimiento")
nulos_mantenimiento = df_mantenimiento.isnull().sum()
porcentaje_nulos_mantenimiento = (nulos_mantenimiento / len(df_mantenimiento)) * 100
pd.DataFrame({"Nulos": nulos_mantenimiento, "Porcentaje": porcentaje_nulos_mantenimiento}).sort_values(by="Porcentaje", ascending=False)


Valores Nulos en df_mantenimiento


,Nulos,Porcentaje
ALIAS,0,0.0
MARCA,0,0.0
MODELO,0,0.0
CONFIGURACION,0,0.0
NO SERIE,0,0.0
DISTRIBUIDOR,0,0.0
SERVICIO,0,0.0
HRMTRO,0,0.0
ACTUAL,0,0.0
TIPO,0,0.0


La columna "O.T." presenta una cantidad considerable de valores nulos, lo que podría indicar que no todos los servicios cuentan con una orden de trabajo registrada.

Asimismo, las columnas "No" y "Editar" presentan valores nulos en todos los registros, por lo que podrían no ser relevantes para el análisis y podrían eliminarse en la etapa limpieza de datos.

In [25]:
print("\nValores Nulos en df_top50")
nulos_top50 = df_top50.isnull().sum()
porcentaje_nulos_top50 = (nulos_top50 / len(df_top50)) * 100
pd.DataFrame({"Nulos": nulos_top50, "Porcentaje": porcentaje_nulos_top50}).sort_values(by="Porcentaje", ascending=False)


Valores Nulos en df_top50


,Nulos,Porcentaje
No.,0,0.0
Unidad,0,0.0
Distribuidor,0,0.0
Marca,0,0.0
Modelo,0,0.0
Estado,0,0.0
Ciudad,0,0.0
Horas,0,0.0


**df_top50:**

No presenta ningún valor nulo en ninguna de sus columnas (0%). Esto indica una excelente calidad de datos en este DataFrame.

## 7. Estadísticas Descriptivas para Variables Numéricas

In [26]:
print("df_population")
df_population.describe()

df_population


,TECHNICAL TYPE,LATITUDE,LONGITUDE,ENGINE HOURS,LATEST COMMUNICATION DATE,BDE PROCESSED AT TIME
count,5.350000e+02,535.000000,535.000000,535.000000,535,535
mean,6.640279e+08,21.029661,-99.858691,410.842991,2026-01-05 16:00:58.042990,2026-02-09 00:27:29.635514
min,5.302014e+08,14.876202,-117.011009,0.000000,2023-08-28 17:10:42,2026-02-08 22:33:33
25%,6.750762e+08,19.097233,-101.725506,2.000000,2026-01-22 02:33:54.500000,2026-02-09 00:33:33
50%,6.960553e+08,20.374104,-99.893349,9.000000,2026-02-07 11:23:59,2026-02-09 00:33:33
75%,6.960553e+08,21.117900,-97.840832,40.000000,2026-02-09 08:51:01.500000,2026-02-09 00:33:33
max,6.986000e+08,32.632122,-86.812210,9762.000000,2026-02-09 12:59:54,2026-02-09 01:33:33
std,5.452089e+07,3.482556,5.072670,1067.722071,NaN,NaN


**Distribución por Marca y Tipo:** Se identificó la frecuencia de marcas como *CASE IH* y *NEW HOLLAND*, así como la frecuencia en  tipos de maquinaria como *TRACTORS* y *LIGHT EQUIPMENT*.

**Segmentación por Horas de Motor:** Se creó una nueva columna *rango_servicio* basándose en las `engine_hours`, creando categorías en rangos como `Nuevo (0-50h)`, `Preventivo 1 (51-300h)`, etc. Podemos notar que la mayoría de las unidades se encuentran en el rango `Nuevo (0-50h)`.

In [27]:
print("df_horas")
df_horas.describe()

df_horas


,Num,IMEI,Latitud,Longitud,2024-01,2024-02,2024-03,2024-04,2024-05,2024-06,2024-07,2024-08,2024-09,2024-10,2024-11,2024-12
count,9405.000000,9.405000e+03,9405.000000,9405.000000,9405.000000,9405.000000,9405.000000,9405.000000,9405.000000,9405.000000,9405.000000,9405.000000,9405.000000,9405.000000,9405.000000,9405.000000
mean,4703.000000,7.116282e+14,21.166979,-100.403035,31.299841,25.007762,24.500266,25.017969,29.839872,25.064115,25.041042,22.594152,20.583626,26.010845,23.495800,22.576289
std,2715.133975,2.339743e+14,3.538277,9.893617,50.632825,42.005026,41.620438,39.602147,45.991285,37.745361,38.100870,36.604668,33.805118,38.717799,33.929141,33.451687
min,1.000000,3.525561e+14,0.000000,-117.018395,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,2352.000000,3.537381e+14,19.204849,-103.319613,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
50%,4703.000000,8.624071e+14,20.521741,-101.538852,2.000000,1.000000,1.000000,2.000000,4.000000,5.000000,4.000000,2.000000,2.000000,5.000000,6.000000,6.000000
75%,7054.000000,8.663921e+14,22.943332,-98.322581,47.000000,35.000000,33.000000,37.000000,45.000000,39.000000,38.000000,32.000000,29.000000,41.000000,37.000000,34.000000
max,9405.000000,8.663921e+14,32.631236,0.000000,397.000000,344.000000,431.000000,295.000000,391.000000,358.000000,285.000000,390.000000,282.000000,367.000000,310.000000,333.000000


**Uso Mensual (df_horas):** Se muestra un resumen estadístico de las horas de uso mensuales, mostrando el comportamiento operativo de las unidades a lo largo del tiempo. También podemos ver que las columnas de los meses están completas.

**Distribución por Marca (df_horas):** Podemos ver que las marcas *NEW HOLLAND AG* y *CASE IH* son las marcas más frecuentes en los datos de telemetría.

**Nulos Críticos (df_horas):** Las columnas `modelo` y `estado` presentan un alto porcentaje de valores nulos (alrededor del 66%), lo que sugiere que esta información no está completa en la fuente de datos.

In [28]:
print("\ndf_unidades")
df_unidades.describe()


df_unidades


,Unnamed: 5,0,0.1,0.2,0.3,0.4,0.5,25.709133,-100.365349,0.6,0.7,0.8,0.9,0.10,0.11,0.12,9,Unnamed: 35
count,10204,10845.000000,10845.000000,10845.000000,10845.00000,10845.000000,10845.000000,10845.000000,10845.000000,10845.000000,10845.000000,10845.000000,10845.000000,10845.000000,10845.000000,10845.000000,10845.000000,10204.000000
mean,2023-04-29 07:28:37.444139,1137.386469,178.007863,5.754265,0.99668,0.712679,4.757584,20.867595,-98.880939,13.574963,5.130764,4.454403,1.099774,1.633144,19.706905,3.007100,5.992900,3.203548
min,1900-01-01 00:00:00,0.000000,0.000000,0.000000,0.00000,0.000000,0.000000,0.000000,-117.017622,-37.960000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,2022-09-14 00:00:00,115.210000,0.360000,0.000000,0.00000,0.000000,0.000000,19.157009,-103.115083,0.000000,0.000000,0.000000,0.000000,0.000000,18.740000,0.000000,4.000000,2.000000
50%,2023-06-19 00:00:00,690.790000,15.960000,3.000000,0.00000,0.000000,2.000000,20.538031,-100.861030,0.000000,0.000000,0.000000,0.000000,0.000000,24.000000,2.000000,7.000000,3.000000
75%,2024-07-29 00:00:00,1644.810000,76.230000,7.000000,1.00000,0.000000,5.000000,22.765297,-98.235734,4.380000,1.440000,7.000000,0.750000,1.440000,24.000000,5.000000,9.000000,4.000000
max,2026-02-09 00:00:00,77300.690000,45518.660000,166.000000,29.00000,29.000000,166.000000,32.624313,0.000000,263.870000,1552.530000,295.000000,100.900000,23.930000,24.000000,9.000000,9.000000,126.000000
std,NaN,1610.859487,1047.319875,10.324985,1.72917,2.036116,10.166067,4.321416,15.440627,31.779551,26.078183,9.941190,2.914772,3.359849,7.705155,2.899911,2.899911,6.330784


`Horometro` tiene una media de 1137.28, lo que indica una distribución extendida de horas de uso entre las unidades. La mediana (690.58) es  menor que la media, lo que sugiere una posible asimetría positiva con algunas unidades de uso muy elevado.
    
`Total` servicios tiene una media de 5.75 con un máximo de 166, lo que indica que algunas unidades han requerido una gran cantidad de intervenciones.

`Latitud` y `Longitud` muestran rangos con valores mínimos de 0.0, lo que sugiere la presencia de datos anómalos o incompletos en algunas entradas (posiblemente coordenadas no registradas o por defecto).

`Combustible` tiene una media de 13.57 y un máximo de 263.87, con un valor mínimo de -37.96. Este valor negativo debe ser investigado o limpiado, ya que las cantidades de combustible no pueden ser negativas.
    
`T. Encendido (hrs)` tiene una media de 1.63 y un máximo de 23.93, lo que indica períodos de operación relativamente cortos para la mayoría de las unidades.

In [29]:
print("\ndf_rutas")
df_rutas.describe()


df_rutas


,Distancia del Dia,Distancia de Ruta
count,3.500000e+01,3.500000e+01
mean,9.999574e+05,9.999574e+05
std,1.492149e+06,1.492149e+06
min,0.000000e+00,0.000000e+00
25%,2.963630e+04,2.963630e+04
50%,9.906280e+04,9.906280e+04
75%,1.613755e+06,1.613755e+06
max,5.306243e+06,5.306243e+06


**Métricas Operativas:** Los descriptores estadísticos muestran los rangos típicos de `distancia_dia`, `distancia_ruta`, `pct_utilizacion`, `pct_visitas_plan` y `pct_visitas_clientes.`

**Divisiones Operativas:** Se identificaron las divisiones o distribuidores únicos presentes en el dataset.

**Información Faltante:** La columna `total_kilos` está completamente vacía *(100% de nulos)*, lo que indica que no se registró esa información, y la columna division tiene un valor nulo.

In [30]:
print("\ndf_mantenimiento")
df_mantenimiento.describe()


df_mantenimiento


,SERVICIO,HRMTRO,ACTUAL
count,23869.000000,23869.000000,23869.000000
mean,1455.448490,1456.988922,2262.148728
std,1394.093725,1397.231893,1710.823506
min,0.000000,0.000000,0.000000
25%,600.000000,560.000000,1064.000000
50%,1200.000000,1160.000000,1870.000000
75%,2100.000000,2060.000000,2974.000000
max,13200.000000,20100.000000,13256.000000


**Métricas de Mantenimiento:** Se obtuvieron descriptores estadísticos para `horas_servicio`, `horometro_objetivo` y `horometro_actual`, se mostraron los *rangos* y la *distribución* de estos valores.

**Estatus y Tipo de Servicio:** Podemos ver que la mayoría de los servicios están en estado *Pendiente* o *CerradaFuera* y la gran mayoría son de tipo *Preventivo*.

**Información Faltante:** Las columnas `id_registro` y `editar` están completamente vacías. `orden_trabajo` tiene un alto porcentaje de nulos *(60.3%)*, lo que nos indica que no todas las órdenes de trabajo están registradas

In [31]:
print("\ndf_top50")
df_top50.describe()


df_top50


,No.,Horas
count,29.000000,29.000000
mean,15.000000,210.673793
std,8.514693,296.151026
min,1.000000,0.040000
25%,8.000000,34.410000
50%,15.000000,136.800000
75%,22.000000,194.980000
max,29.000000,1216.460000


**Estadísticas Numéricas:** Se observaron los descriptores estadísticos de las variables numéricas, nos da una vista general de la distribución de `engine_hours`, `latitude`, `longitude` y otras.

**Distribución por Marca y Tipo:** Se identificó una mayor frecuencia de marcas de tractores como `CASE IH` y `NEW HOLLAND`, y tipos de maquinaria como `TRACTORS` y `LIGHT EQUIPMENT`.

**Segmentación por Horas de Motor:** Se creó una nueva columna `rango_servicio` basándonos en `engine_hours`, categorizando los equipos en rangos como `Nuevo (0-50h)`, `Preventivo 1 (51-300h)`, etc.

Podemos ver que la mayoría de los equipos se encuentran en el rango `Nuevo (0-50h)`.

## 8. Distribución de Variables Categóricas Clave

In [32]:
print("Distribución de 'BRAND' en df_population")
print(df_population['BRAND'].value_counts().head(5))

print("\nDistribución de 'TYPE' en df_population")
print(df_population['TYPE'].value_counts().head(5))

print("\nDistribución de 'Marca' en df_horas")
print(df_horas['Marca'].value_counts().head(5))

print("\nDistribución de 'Modelo' en df_horas")
print(df_horas['Modelo'].value_counts().head(5))

print("\nDistribución de 'Marca' en df_unidades")
print(df_unidades['Marca'].value_counts().head(5))

print("\nDistribución de 'Modelo' en df_unidades")
print(df_unidades['Modelo'].value_counts().head(5))

print("\nDistribución de 'Division' en df_rutas")
print(df_rutas['Division'].value_counts().head(5))

print("\nDistribución de 'MARCA' en df_mantenimiento")
print(df_mantenimiento['MARCA'].value_counts().head(5))

print("\nDistribución de 'ESTATUS' en df_mantenimiento")
print(df_mantenimiento['ESTATUS'].value_counts().head(5))

print("\nDistribución de 'Marca' en df_top50")
print(df_top50['Marca'].value_counts().head(5))

print("\nDistribución de 'Estado' en df_top50")
print(df_top50['Estado'].value_counts().head(5))

Distribución de 'BRAND' en df_population
BRAND
CASE IH                     206
NEW HOLLAND                 176
CASE CONSTRUCTION           145
NEW HOLLAND CONSTRUCTION      8
Name: count, dtype: int64

Distribución de 'TYPE' en df_population
TYPE
TRACTORS                329
LIGHT EQUIPMENT         131
HARVESTING EQUIPMENT     50
HEAVY EQUIPMENT          22
HAYTOOLS                  3
Name: count, dtype: int64

Distribución de 'Marca' en df_horas
Marca
NEW HOLLAND AG    7888
CASE IH           1099
NEW HOLLAND CE     150
0                  142
CASE CE             65
Name: count, dtype: int64

Distribución de 'Modelo' en df_horas
Modelo
7610S HERENCIA    1099
6610S HERENCIA     317
6810               200
TS6.120 ELITE      166
0                  137
Name: count, dtype: int64

Distribución de 'Marca' en df_unidades


KeyError: 'Marca'

Las marcas **NEW HOLLAND AG** y **CASE IH** son las más frecuentes en la mayoría de los DataFrames (`df_population`, `df_horas`, `df_unidades`, `df_mantenimiento`). Esto sugiere que estas son las marcas principales que la empresa gestiona.

Los **TRACTORS** y **LIGHT EQUIPMENT** son los tipos de equipos más comunes en `df_population`, lo que podría indicar el enfoque principal de la flota de maquinaria.

La distribución por `Estado` y `Ciudad` en `df_top50` muestra concentraciones específicas, como **Nuevo León**, lo que podría indicar áreas geográficas clave para las operaciones.

En `df_mantenimiento` podemos observar un alto número de servicios en estado **'Pendiente'** o **'CerradaFuera'**, lo que podría señalar áreas de oportunidad o cuellos de botella en la gestión de servicios.

## 9. Identificación de Columnas para Enlace

Basado en el análisis de las estructuras y diccionarios de datos, las columnas clave identificadas para las uniones y para servir como identificadores únicos entre los diferentes DataFrames son:

*   **IMEI**: Identificador del dispositivo telemático (presente en `df_population`, `df_horas`, `df_unidades`).
*   **VIN / No. Serie**: Identificador único de fábrica del equipo (presente en `df_population` como 'VIN', en `df_unidades` y `df_mantenimiento` como 'NO SERIE' o similar).
*   **Alias / Unidad**: Nombre con el que se reconoce al equipo dentro del sistema (presente en `df_horas`, `df_unidades`, `df_mantenimiento`, `df_top50`).
*   **Marca** y **Modelo**: Atributos comunes que pueden usarse para validación o como parte de claves compuestas (presentes en la mayoría de los DataFrames).
*   **Distribuidor / Division**: Clave jerárquica para relacionar el rendimiento logístico y comercial (`df_horas`, `df_unidades`, `df_rutas`, `df_mantenimiento`, `df_top50`).

## 10. Revisión analítica para definir  hipótesis e identificar patrones

In [ ]:
print("\nTabla con número y porcentaje de unidades Pendiente y Fuera")
porcentaje = (df_mantenimiento['ESTATUS'].value_counts() / len(df_mantenimiento)) * 100

print(df_mantenimiento['ESTATUS'].value_counts(), "\n\nPORCENTAJE" ,porcentaje.round(2))



Tabla con número y porcentaje de unidades Pendiente y Fuera
ESTATUS
Pendiente       12667
CerradaFuera     5612
Cerrada          5037
PorVencer         493
EnProceso          79
Abierta            12
Name: count, dtype: int64 

PORCENTAJE ESTATUS
Pendiente       52.99
CerradaFuera    23.48
Cerrada         21.07
PorVencer        2.06
EnProceso        0.33
Abierta          0.05
Name: count, dtype: float64


Identificamos que más de la mitad de los servicios de mantenimiento están en el estado "Pendiente" (52.99%) y en "CerradaFuera" se identificó una cantidad significativa (23.48%).

Podemos sugerir que existe algún desaprovechamiento operativo o tener alguna desviación en los procedimientos de los mantenimientos. Tener un atraso como estos podrían afectar la fiabilidad y la vida útil de las unidades.

In [ ]:
print("Tabla de antigüedad de servicios pendientes/fuera por rangos de días")
df_mantenimiento_pendientes = df_mantenimiento[df_mantenimiento['ESTATUS'] == 'Pendiente']
df_mantenimiento_fuera = df_mantenimiento[df_mantenimiento['ESTATUS'] == 'Cerrada']

print("df_mantenimiento_pendientes")
print(df_mantenimiento_pendientes['FECHA'].value_counts().head())
print("\ndf_mantenimiento_fuera")
print(df_mantenimiento_fuera['FECHA'].value_counts().head())

Tabla de antigüedad de servicios pendientes/fuera por rangos de días
df_mantenimiento_pendientes
FECHA
2024-05-12 10:08:00    118
2024-04-10 18:47:00     58
21/08/25 10:43          54
2025-11-07 09:25:00     48
18/06/24 10:46          44
Name: count, dtype: int64

df_mantenimiento_fuera
FECHA
18/06/24 10:46         50
18/06/24 10:47         27
2024-04-10 18:47:00    24
2024-05-12 10:08:00    20
2024-05-12 10:14:00    17
Name: count, dtype: int64


El mayor impacto identificado es la inconsistencia de formato en las columnas, es necesario corregirlas para hacer un análisis efectivo.

Las acomulaciones de fechas para pendientes y fuera pueden ser por cargas masivas o mantenimientos que no fueron actualizados correctamente.

In [ ]:
print("Tabla de unidades por nivel de riesgo: Bajo, Medio, Alto")
df_population['SEVERITY LEVEL'].value_counts()

Tabla de unidades por nivel de riesgo: Bajo, Medio, Alto


,count
SEVERITY LEVEL,
Low,485
Warning,56
Critical,44


La mayoría de las unidades están categorizadas con un nivel de severidad bajo (Low), esto es una tendencia positiva, pero el hecho de que sigan unidades con severidad media y alta sigue requeriendo atención.

In [ ]:
df_mantenimiento['delay_vs_service_interval'] = df_mantenimiento['ACTUAL'] - df_mantenimiento['SERVICIO']

status_risk_mapping = {'Cerrada': 0,'PorVencer': 1,'EnProceso': 1,'Abierta': 1,'Pendiente': 2,'CerradaFuera': 2 }


df_mantenimiento['overdue_risk'] = df_mantenimiento['ESTATUS'].map(status_risk_mapping)


rit_agg_df = df_mantenimiento.groupby('ALIAS').agg( avg_overdue_risk=('overdue_risk', 'mean'),
                                      avg_delay_vs_service_interval=('delay_vs_service_interval', 'mean')).reset_index()

In [ ]:
print("Tabla cruzada: Riesgo vs Estatus")
estatus_Unidades = df_mantenimiento['ESTATUS'].value_counts()
riesgo_unidades = df_mantenimiento['overdue_risk'].value_counts()

pd.crosstab(df_mantenimiento['ESTATUS'], df_mantenimiento['overdue_risk'])

Tabla cruzada: Riesgo vs Estatus


overdue_risk,0.0,1.0,2.0
ESTATUS,,,
Abierta,0,12,0
Cerrada,5037,0,0
CerradaFuera,0,0,5612
EnProceso,0,79,0
Pendiente,0,0,12667
PorVencer,0,493,0


La tabla confirma la correcta asignación de niveles de riesgo a cada estado de servicio.
El patrón es que los servicios Pendientes y Cerrados Fuera tienen el mayor riesgo (2.0). Hacer esta categorización nos permite fortalecer la priorización de las unidades.

In [ ]:
print("Tabla cruzada: Riesgo vs Severity Level")

estatus_Unidades = df_mantenimiento['ESTATUS'].value_counts()
riesgo_unidades = df_population['SEVERITY LEVEL'].value_counts()

pd.crosstab(df_mantenimiento['ESTATUS'], df_population['SEVERITY LEVEL'])

Tabla cruzada: Riesgo vs Severity Level


SEVERITY LEVEL,Critical,Low,Warning
ESTATUS,,,
Cerrada,17,137,23
CerradaFuera,9,110,9
EnProceso,0,1,0
Pendiente,18,237,24


Tenemos la relación entre el estatus de los mantenimientos con el nivel de severidad técnica.

Estos datos indican que existen serviciosque no han sido cerrados o que se hicieron fuera de tiempo, eso generaron señales de severidad alta

In [ ]:
#df_mantenimiento

In [ ]:
print("Tabla de frecuencia de servicios por unidad")
df_mantenimiento.groupby(['ALIAS', 'DISTRIBUIDOR']).size().nlargest(10).reset_index(name='Frecuencia de Servicios')

Tabla de frecuencia de servicios por unidad


,ALIAS,DISTRIBUIDOR,Frecuencia de Servicios
0,ATN - NH10382M-001956092,ATN,21
1,ENAGRI - NKJ0B80CANKH30823,ENAGRI,20
2,ARBSA - NH10579M,ARBSA,19
3,ATC - S402743M,ATC,18
4,TEPSA - S402783M,TEPSA,18
5,ATN - 866392068970072,ATN,17
6,ATN - NH10500M,ATN,17
7,TEPSA - NH10125M,TEPSA,17
8,866392068719594,MADISA,16
9,866392069077810,ATN,16


Es importante conocer las unidades con mayor frecuencia de servicios ya que podemos identificar las unidades que requieren atención constante.
Existe el patrón de una concentración de servicios con unidades como ***'ATN - NH10382M-001956092'*** y ***'ENAGRI - NKJ0B80CANKH30823'***. Esto quiere decir que algunas unidades tienen más demanda de mantenimientos.

In [ ]:
print("Tabla de unidades con más servicios acumulados")
df_mantenimiento.groupby('ALIAS')['SERVICIO'].sum().nlargest(10)

Tabla de unidades con más servicios acumulados


,SERVICIO
ALIAS,
ATN - NH10382M-001956092,125700
ATC - S402743M,124500
TEPSA - NH10125M,123600
TEPSA - S402783M,121800
MADISA - S507776M,119100
ATC - NH10163M,111000
ATN - 866392068970072,110700
ARBSA - S507455M,109200
ATC - S402691M,105600


Las horas de servicio realizadas son indicadores para entender que unidades tienen un historial de mantenimiento más extenso.

Podemos identificar un patrón sobre las unidades con mayor servicio acumulado son coherentes con las unidades con mayor frecuencia de servicios, lo que refuerza la idea de que ciertas unidades tienen mayor demanda en cuanto mantenimientos o que tienen un ciclo de vida más largo con varias intervenciones.

In [ ]:
#rit_agg_df


In [ ]:
print("Tabla de probabilidad predictiva de Pendiente/Fuera")
display(rit_agg_df.sort_values(by='avg_overdue_risk', ascending=False).head(10))

Tabla de probabilidad predictiva de Pendiente/Fuera


,ALIAS,avg_overdue_risk,avg_delay_vs_service_interval
6774,tepsa-NKJ0B80CPPKH34446,2.0,1231.0
6772,VRM - NKJ0B80CVPKH34436,2.0,414.0
1,353738100917408,2.0,84.0
2,353738101213682,2.0,449.0
6755,TRIOSA-HFJ182768,2.0,548.0
6754,TRIOSA-HFJ182702,2.0,422.0
6753,TRIOSA-5800066M-ALEJ RMEZ,2.0,1.0
6752,TRIOSA - S204639M-8101302,2.0,1010.0
6751,TRIOSA - S204534M,2.0,433.0
6750,TRIOSA - HFJ184629,2.0,395.0


Una oportunidad de monetización es con las unidades que tienen un ***"avg_overdue_risk"*** y un ***avg_delay_vs_service_interval*** altos, ya que representarían objetivos clave para intervenir y reducir las fugas de .

Priorizarlos y trabajar en estos retrasos y riesgos podrían mejorar el rendimiento de la unidad y reducir costos, así abordamos la oportunidad de monetización a través de un mantenimiento proactivo.

In [ ]:
print("Tabla de score de oportunidad de monetización")
display(rit_agg_df.sort_values(by='avg_delay_vs_service_interval', ascending=False).head(10))

Tabla de score de oportunidad de monetización


,ALIAS,avg_overdue_risk,avg_delay_vs_service_interval
6683,TRACTIESA - S204698M,2.000000,6653.000000
12,353738101261723,2.000000,4761.000000
3043,ATC - 353738102946942,1.857143,4651.714286
4322,ENAGRI - NKJ0B80CANKH30823,2.000000,3929.000000
3483,ATN - NH10382M-001956092,1.666667,3709.285714
3057,ATC - 5800734M-728339207,2.000000,3701.272727
3445,ATN - 866392068970072,1.823529,3349.235294
5603,TEPSA - S402783M,1.444444,3161.333333
3489,ATN - NH10500M,1.411765,3108.823529
5403,TEPSA - NH10125M,2.000000,3065.411765


Las unidades con ***avg_overdue_risk***	y ***avg_delay_vs_service_interval*** altos representan una oportunidad clara de monetización. Priorizar estos mantenimientos  para reducir estos retrasos y riesgos podrían mejorar el rendimiento de las unidades.

Existe el patrón en donde ciertas unidades presentan un alto riesgo de atraso y un gran promedio de retraso en los intervalos de servicio. Estos son los principales objetivos para intervenciones proactivas.